### ISSUE TITLE
indent: function declaration in aligned multi-line arguments

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#4206]
I'm sorry, but it looks like we hit another issue in the indent rule. It should pass, imho. eslint@1.7.2
```
indent: [2,2]
```
```js
functionWithCallback(something,
                     function() {
  var error;
})
```
```
  3:3  error  Expected indentation of 23 space characters but found 2  indent
  4:1  error  Expected indentation of 21 space characters but found 0  indent
```
[Real code example code](https://github.com/nodejs/node/blob/0fc0902c21c9c9ed180e575f34b9b39366d57685/test/parallel/test-promises-unhandled-rejections.js#L238-L240).

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by silverwind** [comment]
Note the bug doesn't trigger unless `function` is the first on its line. This code is fine:

```js
functionWithCallback(something,
                     another, function() {
  var error;
})
```

**Comment by gyandeeps** [comment]
We need to figure out a better way here because we have issues logged asking for opposite behavor to be true also.
https://github.com/eslint/eslint/issues/4174

**Comment by silverwind** [comment]
Maybe be a bit more lenient and allow mulitple correct indendation levels?

**Comment by gyandeeps** [comment]
The rules are very much like 0 or 1 ie either something is wrong or its correct. We have always used options to make sure we can have 2 states to trigger correct behavior.
@eslint/eslint-team thought here?

**Comment by ilyavolodin** [comment]
In all honesty example that @silverwind provided doesn't look right to me. I understand that it's trying to align multi-line arguments based on the position of the first argument, and I guess it's a valid style, it just seems pretty alien to me. For me, unclosed brace or parent means next line should be indented by  one indentation level.

**Comment by Fishrock123** [comment]
> I understand that it's trying to align multi-line arguments based on the position of the first argument, and I guess it's a valid style, it just seems pretty alien to me.

It's what node core has used for a long time. :/

**Comment by trevnorris** [comment]
We may need to rethink this. Historically node has more or less followed the Google Style Guide and this case is covered (under "Passing Anonymous Functions"): https://google.github.io/styleguide/javascriptguide.xml?showone=Code_formatting#Code_formatting

We can see this same style being used in `tls.js`: https://github.com/nodejs/node/blob/v4.2.1/lib/tls.js#L84-L89

Though there is a similar case that violates this in `net.js` (added in 4c150ca): https://github.com/nodejs/node/blob/v4.2.1/lib/net.js#L1133

In the tests I sampled 34 cases of passed anonymous functions. Of those, 23 followed the Google style guide and 11 did not. Of those that did not follow the style all but one commit either occurred back in 2010 or this year. While those that do follow have commits dispersed throughout 2010 to 2015. (below is a list of all files and their associated commits)

This would lead me to believe that the indentation style under question is in fact incorrect and we should move to adhere more closely to the Google style guide.

Followed:
`test-assert.js` (87286cc)
`test-cli-eval.js` (3d22dbf, 83b1dda)
`test-file-write-stream3.js` (87286cc)
`test-fs-write.js` (0665f02)
`test-http-header-read.js` (87286cc)
`test-http-invalidheaderfield.js` (6192c98)
`test-preload.js` (1514b82)
`test-process-binding.js` (962e651)

Not followed:
`test-dgram-bytes-length.js` (f29762f)
`test-dgram-pingpong.js` (0665f02)
`test-dgram-udp4.js` (0665f02)
`test-http-304.js` (093dfaf)
`test-http-timeout.js` (f29762f)
`test-promises-unhandled-rejection.js` (f29762f)

**Comment by nzakas** [comment]
I'm not sure we can support the format being requested. As I'm sure everyone realizes, indentation is a complicated matter, and while we'd like to, it's just not possible to support every possible version that people want.

For reference, here's what the Google style guide shows for passing anonymous functions:

```js
prefix.something.reallyLongFunctionName('whatever', function(a1, a2) {
  if (a1.equals(a2)) {
    someOtherLongFunctionName(a1);
  } else {
    andNowForSomethingCompletelyDifferent(a2.parrot);
  }
});

var names = prefix.something.myExcellentMapFunction(
    verboselyNamedCollectionOfItems,
    function(item) {
      return item.name;
    });
```

We definitely support the first style, and the second style we will support (#4174). If it's possible for Node.js to use one of these two styles, then I'd suggest doing that (after all, you do get autoformatting to fix this with `--fix`).

**Comment by trevnorris** [comment]
Historically node has preferred both those styles. The first if it fits under 80 characters. Otherwise the latter.

**Comment by silverwind** [comment]
I think I agree, we gotta change that style of ours, and I quite like the second option, except I'd put the closing parenthesis on the same level as the opening one:

````js
var names = prefix.something.myExcellentMapFunction(
    verboselyNamedCollectionOfItems,
    function(item) {
      return item.name;
    }
);
````

**Comment by gyandeeps** [comment]
@silverwind The example you gave on top is already supported.

**Comment by targos** [comment]
@gyandeeps What about the following case ?
```js
server.on('listening', function() {
  client.send(message_to_send, 0, message_to_send.length, server_port,
    'localhost', function(err) {
      if (err) {
        console.log('Caught error in client send.');
        throw err;
      }
    }
  );
});
```

I get:
```
  4:7  error  Expected indentation of 4 space characters but found 6  indent
  8:5  error  Expected indentation of 2 space characters but found 4  indent
```

**Comment by gyandeeps** [comment]
@targos Thats exactly what is been discussed in #4174

**Comment by nzakas** [comment]
Unless I'm misreading, it seems like we can close this? @silverwind?

**Comment by gyandeeps** [comment]
As per my understanding, yes (we can close).

**Comment by silverwind** [comment]
@nzakas yeah, we'll try adapting our style.

**Comment by trevnorris** [comment]
@targos Following the Google style guide that would look something like:
```js
server.on('listening', function() {
  client.send(
      message_to_send,
      0,
      message_to_send.length,
      server_port,
      'localhost',
      function(err) {
        if (err) {
          console.log('Caught error in client send.');
          throw err;
        }
      });
});
```
